docs/painless/painless-guide/painless-intro.asciidoc

```diff
@@ -0,0 +1,173 @@
+    "lang": "painless", <2>
+    "params": { <3>
+        "hello_world": "Hello, world!"
+    }
+  }
+}
+----
+[TEST]
+<1> This is a big departure from the previous example. Here we have a
+`Map` of params that are user-defined. We then call the `get` method on
+params to access a value using the `String` key of `hello_world`. There
+are a lot of moving pieces here that we will discuss more about later on.
+<2> We explicitly specify the `Painless` language, but again, it's not
+actually necessary here as `Painless` is the default under most script
+contexts.
+<3> User-defined parameters are a JSON object that is parsed into a `Map`
+made available as a `params` variable within a Painless script. In this
+case, we have added a `String` value of `Hello, world!` with a `String`
+key of `hello_world`. We then access this value within the example script.
+
+===== Hello, world! but concise
+
+Let's make our script a bit more concise using some of the built-in sugar.
+
+[source,console]
+----
+POST /_scripts/painless/_execute
+{
+  "script": {
+    "source": "params['hello_world']",
+    "params": {
+        "hello_world": "Hello, world!"
+    }
+  }
+}
+----
+[TEST]
+
+We have used three pieces of syntactic sugar to replace some verbosity for a
+simple script that is only a single line.
+
+. We first removed the `return` keyword. Painless automatically uses the final
+statement in a script (when able) to produce a return value in a script context
+that requires one.
+. We then removed the semicolon at the end of the statement. Painless does not
+require semicolons for the final statement of a block. (However, it does require
+them in other cases as to remove ambiguity.)
+. Finally, we removed the `get` method and replaced it with braces. Painless
+uses a shortcut specifically for the `Map` type that allows us to use braces
+instead of the lengthier `get` method.
+
+Admittedly, this last example contains a huge amount of information all at once.
+Let's break Painless down into more palatable chunks in the next section.
+
+==== The basics of Painless
+
+A Painless script is structured as one or more statements and optionally
+has one or more user-defined functions at the beginning. A script must always
+have at least one statement.
+
+==== Types Examples
+
+[source,console]
+----
+POST /_scripts/painless/_execute
+{
+  "script": {
+    "source": """
+        int my_int; <1>
+        my_int = 1; <2>
+        double my_double = 2.0; <2>
+        Map my_map; <4>
+        my_map = new HashMap(); <5>
+        my_map.put("my_int", my_int); <6>
+        my_double.put("my_double", my_double);
+        return my_map;
+    """
+  }
+}
+----
+[TEST]
+<1> primitive `int` type declaration
+<2> primitive `int` type assignment with an `int` constant
+<3> primitive `double` type declaration with assignment of a `double` constant
+<4> reference `Map` type declaration
+<5> reference `Map` type assignment with a new instance of a `HashMap`
```

Comment:
It could be useful to explain a bit about the type parameters used by the `HashMap`. Is it `Map<Object, Object>` when translated into java?